src/Database/DisposableFullTextSearch.php

```diff
@@ -0,0 +1,59 @@
+<?php
+/**
+ * @copyright Copyright (C) 2010-2024, the Friendica project
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace Friendica\Database;
+
+/**
+ * Full-text search on a haystack string that isn't present in the database.
+ * The haystack is inserted in a temporary table with a FULLTEXT index, then any number of
+ * matches can be performed on it before the row is deleted when the class instance is destroyed,
+ * either manually or at the end of the script at the latest.
+ */
+class DisposableFullTextSearch
+{
+	private Database $db;
+	private int $identifier;
+
+	public function __construct(Database $database, string $haystack)
+	{
+		$this->db = $database;
+
+		// Maximum value is indicated by the INT UNSIGNED type of the check-full-text-search.pid field
+		$this->identifier = random_int(0, pow(2, 32) - 1);
```

Comment:
Why don't you use the pid?

Reply:
if you create two DisposableFullTextSearch object in the same script execution, the first object destruction would delete both searchtext rows, before the second object destruction is called, leading to unexpected behavior. If each object has it's own random identifier ensures the expected behavior of the specific searchtext deleted only when the corresponding object is destroyed.

Reply:
Since this class is used by us and it isn't some third party class, we can document the behaviour. The problem with a random number is that it is random and that it can happen that the same number can be returned at the same time.

Reply:
Yes, there's always a chance of a collision, but it is cosmically low given the range of the `identifier` field.

I may have found a way to avoid collisions completely using both `getmypid()` and `spl_object_id()`. The entropy will be less, but isn't the point of this field.

Reply:
The chance of a collision is even lower when we just ensure that we never initiate two instances of the class.

Reply:
To handle possible collisions, you could check the return value of the insert command and then could do the random stuff.

Reply:
The return value will be `true` because there's no auto increment `id` field (nor should there be).

> The chance of a collision is even lower when we just ensure that we never initiate two instances of the class.

There's no need to make it into a Singleton class. The current randomness is enough to prevent collisions at the scale this class is meant to be used. Of course, if the `identifier` was converted into a string, it would be possible to aggregate the server hostname, the script PID and the object ID in a way that would be unique for each instance of this class even on a node with multiple frontend servers and a single database backend.

Reply:
When you remove the `Database::INSERT_UPDATE` in the insert command, then it will fail on a duplicate entry. And then the return value of the `Insert` command will be `false`. So then you could try again with a different value. Or you could add an additional primary key that contains the timestamp, since you cannot have the same timestamp with the same pid.

Reply:
On a distributed system (multiple frontends with a single database backend), you can definitely have the same pid and the same timestamp.

And the timestamp returned by microtime would either need to be a string, or a large integer if multiplied by 10000. In either case there are more reliable ways to either generate entropy, or ensure uniqueness.

Reply:
We currently don't support distributed systems. When we do so, we will have to add an additional factor also in the worker queue, like some "host id".

This class is called from two places. And it is unlikely that it will be called at much more places. Also it is just some internally used class. This means that we can be sure that there won't be multiple instances at the same time.

Reply:
I implemented a loop with the insert because I don't want to change the DB schema for this.